### ISSUE TITLE
`peer_streams_left_bidi()` value never increases when streams finish

### ISSUE THREAD

**Issue by tegefaulkes** [cloudflare/quiche#1883]
Within our program we kept running into a failure to initiate a stream due to hitting the stream limit after using a connection for a while. At first I thought we had a stream leak where we didn't take the proper steps to clean up the streams properly. However after a bunch of debugging I found as best as I can tell we we're cleaning them up properly. Despite this the `peer_streams_left_bidi()` function was showing an ever decreasing number of available streams.

Looking through  the rust docs I found https://docs.rs/quiche/latest/quiche/struct.Config.html#method.set_initial_max_streams_bidi suggesting that the limit applies to active streams. However digging through the code https://github.com/cloudflare/quiche/blob/0570ab83cc5e46dc7b877765a6c0d7c4a44dd885/quiche/src/stream/mod.rs#L247C1-L259C23 I can only see the value being set as the max of the current value or the stream sequence. Does this value ever decrease when a stream has ended?

So in summary
- The docs state that the limit applies to active streams.
- From testing and digging through the code, the counter `local_opened_streams_bidi` never actually decreases so it seems the limit applies to all created streams for the life of the connection.

What is the intended behaviour here?

**Comment by LPardue** [comment]
Do you control both client and server? What role are you seeing peer stream limit getting reached?

In QUIC, the peer needs to issue additonal stream credits via MAX_STREAMS.

**Comment by tegefaulkes** [comment]
Yes, we control both the client and the server. Both of them are running quiche.

**Comment by LPardue** [comment]
Thanks for the clarification. I would expect that quiche attempts to maintain a stream concurrency limit of what is configured in set_initial_max_streams_bidi.

For example, if a server offers 100 initial streams, it will emit a MAX_STREAMS frame when even 1 just stream fully executes it's lifecycle. I.e. the client has marked the stream with a FIN  or sent a RESET_STREAM, and the server has marked the stream with a FIN or RESET_STREAM.

It's hard to know without seeing the code or logs whether you're fully doing this.

**Comment by tegefaulkes** [comment]
Ok, So i dug through the code some more and yes, I can see the `local_max_streams_bidi` getting increased when the stream gets collected. That suggests that my streams are not being cleaned up properly.

During debugging I can confirm that both sides of the stream get to a state where `stream_finished()` returns true. That shows the `FIN` or `RESET_STREAM` on the readable. And the writable returns `InvalidState()` which I think indicates that the stream has been cleaned up? Its not clear from the docs what indicates that the stream has been fully cleaned up.

The repository is at https://github.com/MatrixAI/js-quic if you want to look over it. But it will be tricky to provide a clear and self contained example of what's happening.

**Comment by LPardue** [comment]
Is it possible at all for you to enable qlog and to share an example log when this occurs?

**Comment by tegefaulkes** [comment]
Sure, I'll look into it.

**Comment by LPardue** [comment]
You might also want to look into the `should_update_max_streams_bidi()` function, since that controls whether a MAX_STREAMS frame is emitted during `send()` operations

**Comment by tegefaulkes** [comment]
Thanks for the help. After further digging I found that the limit is replenishing fine, it just happens in chunks larger than the number of streams I was testing with.

Here's a log of the remaining limit and the delta. Each line is printed after a stream is created and finishes sending and reading.
```
WARN:QUICStream Test:peerStreamsLeftBidi() 99, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 98, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 97, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 96, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 95, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 94, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 93, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 92, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 91, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 90, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 89, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 88, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 87, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 86, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 85, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 84, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 83, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 82, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 81, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 80, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 79, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 78, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 77, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 76, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 75, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 74, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 73, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 72, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 71, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 70, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 69, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 68, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 67, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 66, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 99, -33
WARN:QUICStream Test:peerStreamsLeftBidi() 98, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 97, 1
WARN:QUICStream Test:peerStreamsLeftBidi() 96, 1
```
So I think in our case, the problem was caused by running through the limit of 100 before the limit could be replenished during periods of high activity.

I'm going to close out this issue. Thanks again.

**Comment by LPardue** [comment]
All good. Providing more control of stream limits to apps is something we might do further down the line.
But not something being actively planned or worked on.

In the meantime, you might want to consider of you can afford the resource cost o provide a larger concurrency limit to soak up bursts.
